test/performance/odp_crc.c

```diff
@@ -0,0 +1,238 @@
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+/* Command line options */
+typedef struct {
+	uint32_t time;
+} options_t;
+
+static options_t options;
+static const options_t options_def = {
+	.time = 1000,
+};
+
+static void print_usage(void)
+{
+	printf("\n"
+	       "CRC performance test\n"
+	       "\n"
+	       "Usage: odp_crc_perf [options]\n"
+	       "\n"
+	       "  -t, --time    Approximate time (ms) for each measurement (default %u)\n"
+	       "  -h, --help    This help\n"
+	       "\n",
+	       options_def.time);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+	int opt;
+	int long_index;
+	int ret = 0;
+
+	static const struct option longopts[] = {
+		{ "time", required_argument, NULL, 't' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+t:h";
+
+	options = options_def;
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			options.time = atol(optarg);
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (options.time < 1) {
+		ODPH_ERR("Invalid time: %" PRIu32 "\n", options.time);
+		return -1;
+	}
+
+	return ret;
+}
+
+static uint8_t *buf, *data;
+static const int bufsize = 32 * 1024;
+
+static void prepare_data(void)
+{
+	uint64_t seed = 1;
+	const unsigned long page = 4096;
+
+	/* One extra page for alignment, another one for offset. */
+	buf = (uint8_t *)malloc(bufsize + page * 2);
+	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
+
+	if (odp_random_test_data(data, bufsize, &seed) != bufsize) {
+		ODPH_ERR("odp_random_test_data() failed.\n");
+		exit(EXIT_FAILURE);
+	}
+}
+
+static void free_data(void)
+{
+	free(buf);
+}
+
+typedef uint32_t (*test_fn_t)(const void *data, uint32_t data_len,
+			      uint32_t init_val);
+
+static uint64_t measure(test_fn_t fn, uint32_t size, uint32_t offset,
+			uint32_t num)
+{
+	void *p = data + offset;
+	volatile uint32_t crc = 1;
+	odp_time_t start = odp_time_local();
+
+	for (uint32_t i = 0; i < num; i++) {
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+		crc ^= fn(p, size, crc);
+	}
+
+	return odp_time_diff_ns(odp_time_local(), start);
+}
+
+static int test_with_params(test_fn_t fn, uint32_t size, uint32_t offset)
+{
+	uint64_t nsec;
+	uint64_t num = 1;
+	double mb, seconds;
+
+	/* Search for a good number of loops. */
+	while (1) {
+		nsec = measure(fn, size, offset, num);
+
+		if (nsec / 1000000 > options.time / 2)
+			break;
+
+		num *= 2;
+	}
+
+	num = (double)num * (double)options.time / 1000.0 /
+	      ((double)nsec / (double)ODP_TIME_SEC_IN_NS);
```

Comment:
This time measurement makes test runs not repeatable. The same parameters may result different number of rounds (and data size).

Instead of time, user could pass test length in MB. That would tell number of rounds for each data size, and have about the same execution time on each case. The default length could be pretty small (e.g. 1MB -> 32x32kB), so that user (or CI) don't think that test hang when CRC implementation is slow (default SW).

Reply:
Done in v2, although this means that the test doesn't produce any useful results with the default parameters. And also that for example on x86 the crc32 test takes over 15 times longer than the crc32c test, which seems completely impractical. Should I do something about that?

Reply:
The default size (-s) could be e.g. 16kB, so that it surely fits into L1 cache. With that -m could be 32MB, which results about 0.1sec in my system with the SW based CRC. Which CRC to test could be also an option, so that user can pick values based on the algorithm under test. Default selection could be both (so that CI runs both).

Now, when there's single size in test, -m option could be simply number of rounds (-r). It's may be more intuitive what is tested: e.g. 10000x16kB.

Reply:
Done in v3. The default is now 10000x16KB.